Adding `anon` v2 support to the dockerfile

thesuhas opened the pull request:
## Problem

Removed `anon` v1 support as described here: https://github.com/neondatabase/cloud/issues/22663

Adding `anon` v2 support to re-introduce the `pg_anon` extension. 
Related Issues: https://github.com/neondatabase/cloud/issues/20456


## Summary of changes

Adding `anon` v2 support by building it in the dockerfile

thesuhas commented:
> > [neondatabase/cloud#22663 (comment)](https://github.com/neondatabase/cloud/issues/22663#issuecomment-2734001078)
> 
> Why are you adding it back in?

The idea was to remove `v1` support and then reintroduce `v2` support as detailed here: https://github.com/neondatabase/cloud/issues/20456

thesuhas commented on compute/patches/anon_v2.patch:
```diff
@@ -0,0 +1,10425 @@
++
++-- NEON Patches
++
++GRANT EXECUTE ON FUNCTION pg_config to neon_superuser;
```
I will remove the `pg_config` access. 

I have not tried the helper function with superuser security definer. I will try that and update

thesuhas replied:
That works! I have updated the patch

ololobus commented on compute/patches/anon_v2.patch:
```diff
@@ -0,0 +1,78 @@
++RETURNS BOOLEAN AS
++$$
++BEGIN
++  RETURN anon.load_csv(dest_table, csv_file);
```
> I'd rather not give everyone access to superusers copying arbitrary files into their tables.

Yeah, +1, that's actually what I meant in my suggestion. Have a wrapper function that copies only these specific files, put a security definer on this func and call it from the init. That way, the only action users can do with it is to copy exactly the same files into exactly the same tables

thesuhas commented on compute/patches/anon_v2.patch:
```diff
@@ -0,0 +1,78 @@
++
++-- NEON Patches
++
++GRANT EXECUTE ON FUNCTION pg_config to neon_superuser;
```
But, from a user's perspective, all the [anon documentation](https://postgresql-anonymizer.readthedocs.io/en/stable/static_masking/) points to using `anon.init()` and not `anon.init('the-datapath')`. 
Wouldn't this be confusing? Additionally, the user may not be aware of the default `datapath` to be used, to call `anon.init('the-datapath')` with.

thesuhas replied:
I did not think of that, thanks!

thesuhas commented on vendor/revisions.json:
I'm not entirely sure either, I've compared the values to `main` and the commit hashes are exactly the same. As it's the same value, shouldn't this not show up as a diff?

thesuhas commented on compute/patches/anon_v2.patch:
```diff
@@ -0,0 +1,108 @@
+ SECURITY LABEL FOR anon ON FUNCTION anon.load_csv IS 'UNTRUSTED';
+ 
++CREATE OR REPLACE FUNCTION anon.load_csv_superuser(
++  datapath TEXT
```
But, how else would you provide users the functionality to load their own [custom fake data](https://postgresql-anonymizer.readthedocs.io/en/stable/custom_fake_data/)? Or, do you propose removing that and enforcing the usage of fake data we provide (the default fake data provided by the `anon` extension)?

ololobus replied:
Users cannot upload files to compute, so it won't work anyway. Is neon_superuser able to access these tables with fake data? If it can, then users can just copy extra data with with SQL or client-side `/copy`

thesuhas replied:
`neon_superuser` can access these tables. I'll modify this function to hard-code the datapath

thesuhas commented:
I've removed the `anon.init('datapath')` function in its entirety, and now users cannot call it explicitly. I've also updated the `load_fake_data` function (previously `load_csv_superuser`) to hardcode the datapath instead of a user specified one. Users should not be able to load data from the file system anymore.
I've also verified that if users want to load their own fake data, they can insert rows into the tables that `anon` creates consisting of fake data. 
This image also works with the script provided by workflow.
